models/staging/stg_soroban_contract_code.sql

```diff
@@ -0,0 +1,191 @@
+            AND timestamp(cc.closed_at) >= timestamp_sub(timestamp('{{ var("airflow_start_timestamp") }}'), interval 4 hour)
+    {% endif %}
+    )
+
+    , ttl_data as (
+        select
+            ttl.key_hash
+            , ttl.closed_at as ttl_closed_at
+            , ttl.live_until_ledger_seq
+            , ttl.deleted as ttl_deleted
+            , row_number() over (partition by ttl.key_hash order by ttl.closed_at desc) as row_num
+        from
+            {{ source('crypto_stellar', 'ttl') }} as ttl
+
+        -- If the target table is empty, pull all data for first-time run
+        {% if is_empty_target_table %}
+        -- No WHERE condition here, as we want to pull all records
+
+        {% elif var('recovery', 'False') == 'True' %}
+        -- If recovery is enabled, apply the appropriate recovery filter
+
+        -- Single day recovery
+        {% if var('recovery_type', 'SingleDay') == 'SingleDay' %}
+            WHERE DATE(ttl.closed_at) = '{{ var("recovery_date") }}'
+
+        -- Range recovery
+        {% elif var('recovery_type', 'Range') == 'Range' %}
+            WHERE DATE(ttl.closed_at) BETWEEN '{{ var("recovery_start_day") }}' AND '{{ var("recovery_end_day") }}'
+
+        -- Full recovery
+        {% elif var('recovery_type', 'Full') == 'Full' %}
+            -- No WHERE condition for full recovery, as we want all data
+        {% endif %}
+
+    -- Regular cadence filtering (2-hour forward window, 4-hour backward window)
+    {% else %}
+        WHERE
+            timestamp(ttl.closed_at) < timestamp_add(timestamp('{{ var("airflow_start_timestamp") }}'), interval 2 hour)
+            AND timestamp(ttl.closed_at) >= timestamp_sub(timestamp('{{ var("airflow_start_timestamp") }}'), interval 4 hour)
+    {% endif %}
+    )
+
+    , source_data as (
+        select
+            cc.contract_code_hash
+            , cc.last_modified_ledger
+            , cc.ledger_entry_change
+            , cc.ledger_sequence
+            , cc.ledger_key_hash
+            , cc.n_instructions
+            , cc.n_functions
+            , cc.n_globals
+            , cc.n_table_entries
+            , cc.n_types
+            , cc.n_data_segments
+            , cc.n_elem_segments
+            , cc.n_imports
+            , cc.n_exports
+            , cc.n_data_segment_bytes
+            , cc.deleted
+            , cc.closed_at
+            , ttl.key_hash
+            , ttl.ttl_closed_at
+            , ttl.live_until_ledger_seq
+            , ttl.ttl_deleted
+            , cc.batch_insert_ts
+            , cc.airflow_start_ts
+            , cc.batch_id
+            , cc.batch_run_date
+        from
+            contract_code as cc
+        left join
+            ttl_data as ttl
+            on
+            cc.ledger_key_hash = ttl.key_hash
+        where
+            cc.row_num = 1  -- Only the latest record per ledger_key_hash from contract_code
```

Comment:
I don't think using `row_number` and taking the latest record is the correct thing to do for an SCD type 2 table. `lead` would probably be more fitting as we want to record every change

Reply:
I opted to go for row_num nd get the latest for each batch run window (2 hrs)  instead of recording each and every change based on this conversation and some profiling
https://stellarfoundation.slack.com/archives/C073MNCNVQQ/p1726585866116489

SCD can record each and every materialistic change that are meaningful to the end users.

Ex: 
`select
*
from
`crypto-stellar.crypto_stellar.contract_code`
where contract_code_hash='4b9316721487281d8201e1c6044544400f120253487971e339eb23a465516935'
order by ledger_key_hash,closed_at
;`

It has ledger entry related values change but the columns that Users would be interested (n_ fields) only changed twice across 10 records. That is also the reason I went with this approach capturing the latest change per each batch window run (which would be 2 hours in this case)

Reply:
Ideally we'd have both a dim table that records every ledger state change as well as a dim table with daily changes. But yeah for now compacting makes sense. 

I would recommend compacting to daily if the most likely use case is daily+ aggregates